File: bench/benches/sortby.exs
Code.require_file("../lib/bench.exs", __DIR__)
# sortby: N (key, idx) records ordered by a user comparator on a derived key (key%4096, then key);
# checksum = rolling hash of the idx sequence.
defmodule Sortby do
  def gen_fwd(n) do
    {list, _} = Enum.map_reduce(0..(n - 1), 1, fn i, x ->
      nx = rem(16807 * x, 2147483647)
      {{nx, i}, nx}
    end)
    list
  end
  def less({ka, _}, {kb, _}) do
    ma = rem(ka, 4096); mb = rem(kb, 4096)
    ma < mb or (ma == mb and ka <= kb)
  end
  def hsh(list), do: Enum.reduce(list, 0, fn {_, i}, h -> rem(h * 31 + i, 1000000007) end)
end
Bench.run("sortby", fn -> Sortby.hsh(Enum.sort(Sortby.gen_fwd(5000), &Sortby.less/2)) end)

File: bench/benches/sortby.py
import sys, os
sys.path.insert(0, os.path.join(os.path.dirname(__file__), "..", "lib"))
from bench import bench
# sortby: N (key, idx) records ordered by a user key function (key%4096, then key);
# checksum = rolling hash of the idx sequence.
N = 5000
def work():
    x = 1
    data = []
    for i in range(N):
        x = (16807 * x) % 2147483647
        data.append((x, i))
    data.sort(key=lambda r: (r[0] % 4096, r[0]))
    h = 0
    for _, i in data:
        h = (h * 31 + i) % 1000000007
    return h
bench("sortby", work)
